### PR TITLE
Rework SSL Handshake Hooks and add tls_hooks tests.

### PR DESCRIPTION
**Pull request by shinrich**
Reworking how the preaccept, SNI, and cert callbacks are processed to be more consistent.  Where there were two state variables there is now one state variable.  I need to update the documentation to encode the state machine diagram in my notebook.

This is in response to problems @calavera saw with the cert callback in 7.1.  This PR should be a more comprehensive solution than the solution in PR #2152.   There was a little fixing up in the actually buffering handshake processing that was changed by PR #1941, but I think that most of the problems @calavera was seeing was due to faulty handshake hook state processing. 

This PR also includes a set of tls_hook tests which should catch regressions in this error in the future.

I ran autopep8.   It looks like it "fixed up" tests that I'm not messing with.

### CODE REVIEW

**Comment by zwoop** [comment]
Wow, this is a really large change, are we really sure / comfortable this is needed this late in the 7.1.0 release cycle?

**Comment by bryancall** [comment]
I don't feel comfortable with adding this into the 7.1.0 release without more testing.  I suggest we make a 7.1.0 release without this change and after we have tested it more roll out a 7.1.1 release with this change in it.

**Comment by shinrich** [comment]
Looks like I need to figure out why the test plugin doesn't build on FC25.

**Comment by calavera** [comment]
I'm testing this change rebased on top of 7.1.0. The first time I perform a handshake against a servername, the hook doesn't preserve the vconn mutex, even when we don't reenable the connection:

```
david@cdn-test-aws-sfo:/opt/ts$ curl  -s -vv --resolve "www.netlify.com:443:127.0.0.1" https://www.netlify.com
* Added www.netlify.com:443:127.0.0.1 to DNS cache
* Rebuilt URL to: https://www.netlify.com/
* Hostname was found in DNS cache
*   Trying 127.0.0.1...
* Connected to www.netlify.com (127.0.0.1) port 443 (#0)
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
* SSLv3, TLS alert, Server hello (2):
* error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure
* Closing connection 0
david@cdn-test-aws-sfo:/opt/ts$ sudo tail -f /opt/ts/var/log/trafficserver/traffic.out               
[Jun 28 21:31:57.546] Server {0x2ac3bfd70700} DIAG: (netlify_sni) SSL Servername
[Jun 28 21:31:57.546] Server {0x2ac3bfd70700} DIAG: (netlify_sni) Got SNI servername: www.netlify.com
[Jun 28 21:31:57.546] Server {0x2ac3bfd70700} DIAG: (netlify_sni) Lookup for www.netlify.com
[Jun 28 21:31:57.547] Server {0x2ac3c5bd5700} DIAG: (netlify_sni) load_cert
[Jun 28 21:31:57.547] Server {0x2ac3c5bd5700} DIAG: (netlify_sni) Loading cert www.netlify.com
[Jun 28 21:31:57.547] Server {0x2ac3c5bd5700} DIAG: (netlify_sni) Need HTTP request www.netlify.com
[Jun 28 21:31:57.547] Server {0x2ac3c5bd5700} DIAG: (netlify_sni) Doing HTTP request for https://api.netlify.com/certificates/www.netlify.com/resolve
[Jun 28 21:31:57.548] Server {0x2ac3bfd70700} DEBUG: <SSLUtils.cc:1418 (ssl_callback_info)> (ssl) ssl_callback_info ssl: 0x2ac3c8009220 where: 16392 ret: 552
[Jun 28 21:31:57.548] Server {0x2ac3bfd70700} DEBUG: <SSLUtils.cc:1418 (ssl_callback_info)> (ssl) ssl_callback_info ssl: 0x2ac3c8009220 where: 8194 ret: -1
[Jun 28 21:31:57.548] Server {0x2ac3bfd70700} DEBUG: <SSLUtils.cc:1418 (ssl_callback_info)> (ssl) ssl_callback_info ssl: 0x2ac3c8009220 where: 8194 ret: -1
[Jun 28 21:31:57.548] Server {0x2ac3bfd70700} DEBUG: <SSLUtils.cc:2208 (SSLAccept)> (ssl.error.accept) SSL accept returned -1, ssl_error=1, ERR_get_error=336109761 (error:1408A0C1:SSL routines:ssl3_get_client_hello:no shared cipher)
[Jun 28 21:31:57.548] Server {0x2ac3bfd70700} DEBUG: <SSLNetVConnection.cc:1094 (sslServerHandShakeEvent)> (ssl-diag) SSL::47020225529600:error:1408A0C1:SSL routines:ssl3_get_client_hello:no shared cipher:s3_srvr.c:1418: peer address is 127.0.0.1
[Jun 28 21:31:57.548] Server {0x2ac3bfd70700} DEBUG: <SSLNetVConnection.cc:1094 (sslServerHandShakeEvent)> (ssl-diag) SSL handshake error: SSL_ERROR_SSL (1), errno=0
[Jun 28 21:31:57.548] Server {0x2ac3bfd70700} DEBUG: <SSLNetVConnection.cc:1213 (sslServerHandShakeEvent)> (ssl-diag) SSLNetVConnection::sslServerHandShakeEvent, SSL_ERROR_SSL errno=0
[Jun 28 21:31:57.696] Server {0x2ac3bfe72700} DIAG: (netlify_sni) AsyncHttpFetch completed 10000
[Jun 28 21:31:57.696] Server {0x2ac3bfe72700} DIAG: (netlify_sni) Parsing cert JSON
[Jun 28 21:31:57.696] Server {0x2ac3bfe72700} DIAG: (netlify_sni) Creating SSL context
[Jun 28 21:31:57.696] Server {0x2ac3bfe72700} DEBUG: <SSLUtils.cc:1486 (SSLInitServerContext)> (ssl.session_cache) ssl context=0x2ac3cc024a60: using session cache options, enabled=2, size=102400, num_buckets=256, skip_on_contention=0, timeout=0, auto_clear=1
[Jun 28 21:31:57.696] Server {0x2ac3bfe72700} DEBUG: <SSLUtils.cc:1508 (SSLInitServerContext)> (ssl.session_cache) enabling SSL session cache with ATS implementation
[Jun 28 21:31:57.696] Server {0x2ac3bfe72700} DEBUG: <SSLUtils.cc:1522 (SSLInitServerContext)> (ssl) enabling SSL_MODE_RELEASE_BUFFERS
[Jun 28 21:31:57.696] Server {0x2ac3bfe72700} DEBUG: <SSLUtils.cc:1758 (SSLInitServerContext)> (ssl) ssl ocsp stapling is disabled
[Jun 28 21:31:57.696] Server {0x2ac3bfe72700} DIAG: (netlify_sni) Set certificate
[Jun 28 21:31:57.696] Server {0x2ac3bfe72700} DIAG: (netlify_sni) Set private key
[Jun 28 21:31:57.697] Server {0x2ac3bfe72700} DIAG: (netlify_sni) Adding ca cert
[Jun 28 21:31:57.697] Server {0x2ac3bfe72700} DIAG: (netlify_sni) Setting ssl ctx for www.netlify.com
[Jun 28 21:31:57.697] Server {0x2ac3bfe72700} DIAG: (netlify_sni) Reenable vconn
[Jun 28 21:35:32.646] Server {0x2ac3bfd70700} DEBUG: <SSLNextProtocolAccept.cc:127 (mainEvent)> (ssl) [SSLNextProtocolAccept:mainEvent] event 202 netvc 0x2ac3fc01bca0
```

We have this code in our plugin, inside the handleAsyncComplete callback for an AsyncHttp request:

```
  void handleAsyncComplete(AsyncHttpFetch &async_http_fetch)
  {
    ...
    cleanup:
    TSDebug(SNI_TAG_NAME, "Reenable vconn");
    TSMutex mutex = TSContMutexGet(this->vconn);
    if (mutex) {
      TSMutexLock(mutex);
      TSDebug(SNI_TAG_NAME, "VConn mutex locked");
      if (ssl_ctx) {
        TSSslConnection sslobj = TSVConnSSLConnectionGet(vconn);
        if (sslobj) {
          ssl = reinterpret_cast<SSL *>(sslobj);
          SSL_set_SSL_CTX(ssl, ssl_ctx);
          TSDebug(SNI_TAG_NAME, "SSL context set");
        }
      }
      TSVConnReenable(vconn);
      TSDebug(SNI_TAG_NAME, "VConn reenabled");
      TSMutexUnlock(mutex);
    }
```

The execution never gets into the `if (mutex)`.

When I test our plugin rebased on master I get a different error, not caused by this change, but by a different change in ATS. It looks like there responses from my AsyncHttp get truncated in advance. Which prevents me all together from testing this change

**Comment by calavera** [comment]
Totally unrelated, but this is what's blocking me from testing this change in master: https://github.com/apache/trafficserver/issues/2210

**Comment by shinrich** [comment]
Pushed a fix that should make the delayed responses work properly.  Setting the invoked state got messed up at some point in the preparation of the PR.  Extended the test delay to 2 seconds to make sure that hooks don't trigger too soon.  Should probably add a time out test.

File: tests/tools/plugins/ssl_hook_test.cc [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+  TSContDataSet(cb, ssl_vc);
+
+  // Schedule to reenable in a bit
+  TSContSchedule(cb, 2000, TS_THREAD_POOL_NET);
+
+  return TS_SUCCESS;
+}
+
+int
+CB_SNI(TSCont cont, TSEvent event, void *edata)
```
Comment (persiaAziz) [inline tests/tools/plugins/ssl_hook_test.cc]: The crash I was seeing in ubuntu was due to SSL_get_servername((SSL*)ssl_vc, ). The following code worked for me
```
    int

    CB_SNI(TSCont cont, TSEvent event, void *edata)

    {

      TSVConn ssl_vc         = reinterpret_cast<TSVConn>(edata);

    TSSslConnection ssl_conn = TSVConnSSLConnectionGet(ssl_vc);

      auto *ssl                = reinterpret_cast<SSL *>(ssl_conn);

      char const *servername          = SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);

     

    if(ssl_vc){

      //const char *servername = SSL_get_servername((SSL *)ssl_vc, TLSEXT_NAMETYPE_host_name);

     TSDebug(PN, "reallyyyyy   SNI callback - SNI is %p",edata);

      int count = reinterpret_cast<intptr_t>(TSContDataGet(cont));

      TSDebug(PN, "SNI callback %d %p - SNI is %s", count, ssl_vc, servername);

      // All done, reactivate things

      TSVConnReenable(ssl_vc);

    }

      return TS_SUCCESS;

    }
```

**Comment by shinrich** [comment]
[approve ci autest]

**Comment by calavera** [comment]
I rebased these changes on top of 7.1.x and all my tests pass now.

🎉 🎊 🎉

**Comment by shinrich** [comment]
[ci approve autest]

**Comment by shinrich** [comment]
[approve ci autest]

**Comment by calavera** [comment]
@bryancall it's not my decision to make, but from my point of view, the code in the 7.1.x branch is very broken for a production use of ssl hooks. The changes that really impact this behavior are easier to understand now.

I deployed this patch to one of our production PoPs for 2 hours, I got about 5K successful requests that exercised this logic. I only reverted the deploy because it's Friday and I don't want to be paying attention to the log for the rest of the long weekend 🤣

**Comment by zwoop** [comment]
I've added this to the 7.x project, as a 7.1.1 candidate.

**Comment by zwoop** [comment]
@shinrich How safe is this to land, for the use cases when people are not using this hook?

**Comment by shinrich** [comment]
@zwoop I think this should not affect folks not relying on the SSL hooks.  I'm still battling with trying to get the autest to run.  The test plugin runs on a Fedora 25 VM and the Fedora 25 docker image, but tsxs fails when running via jenkins.  Will try to add more shell output in the tsxs to debug.  Downloading the failing docker image is not working.

**Comment by zwoop** [comment]
Lets keep this as an 8.0.0 milestone for now, to make sure we know what goes where (since we can not / should not merge this to 7.1.x quite yet). I've already added this to the 7.x project, under the 7.1.0 candidates, so we won't forget it.

**Comment by shinrich** [comment]
Updated the branch to remove spacing changes on un-related tests.  

I am now running this branch on our version of 7.1.x.  Our plugins are only exercising the servername hook.  So far so good.

**Comment by bryancall** [comment]
@calavera We are marking it for the 7.1.1 release.

**Comment by calavera** [comment]
Thanks @bryancall.

**Comment by zwoop** [comment]
@shinrich This does not cherry-pick cleanly to 7.1.x branch, can you please make a PR against this branch?

**Pull request merged**